dm/dm/worker/subtask.go

```diff
@@ -897,24 +886,47 @@ func updateTaskMetric(task, sourceID string, stage pb.Stage, workerName string)
 }
 
 func (st *SubTask) GetValidatorError(errState pb.ValidateErrorState) []*pb.ValidationError {
-	st.RLock()
-	defer st.RUnlock()
-	if st.validator != nil && st.validator.Started() {
-		return st.validator.GetValidatorError(errState)
+	validator := st.getValidator()
+	if validator != nil && validator.Started() {
+		return validator.GetValidatorError(errState)
 	}
 	st.l.Warn("validator not start")
 	// todo: should it inform the user of this error
 	return []*pb.ValidationError{}
 }
 
 func (st *SubTask) OperateValidatorError(op pb.ValidationErrOp, errID uint64, isAll bool) error {
-	st.RLock()
-	defer st.RUnlock()
-	if st.validator != nil && st.validator.Started() {
-		return st.validator.OperateValidatorError(op, errID, isAll)
+	validator := st.getValidator()
+	if validator != nil && validator.Started() {
+		return validator.OperateValidatorError(op, errID, isAll)
 	}
 	st.l.Warn("validator not start")
 	// todo: should it inform the user of this error
 	// silently exists
 	return nil
 }
+
+func (st *SubTask) getValidator() *syncer.DataValidator {
+	st.RLock()
+	defer st.RUnlock()
+	return st.validator
+}
+
+func (st *SubTask) GetValidatorStatus() (*pb.ValidationStatus, error) {
+	validator := st.getValidator()
+	// todo: should be able to get status even validator is stopped
+	if validator == nil || !validator.Started() {
```

Comment:
ditto

Reply:
see the todo, can only run when it's started now